### ISSUE TITLE
Debug message not clear in InMemoryK1Cache

### ISSUE THREAD

**Issue by straumat** [theborakompanioni/bitcoin-spring-boot-starter#109]
**Expected behavior**

I'm trying to implement your spring boot starter; as I'm having issues, I added logs:

Just after starting, I go the login page and try to log with my wallet :
```bash
2024-01-22T21:23:14.690+01:00  INFO 5018 --- [  restartedMain] o.s.b.w.embedded.tomcat.TomcatWebServer  : Tomcat started on port 8080 (http) with context path ''
2024-01-22T21:23:14.701+01:00  INFO 5018 --- [  restartedMain] org.royllo.explorer.web.WebApplication   : Started WebApplication in 8.914 seconds (process running for 9.551)
2024-01-22T21:23:14.710+01:00  INFO 5018 --- [oyllo-process-1] r.e.c.s.s.StatisticServiceImplementation : Clearing global statistics cache.
2024-01-22T21:23:17.534+01:00  INFO 5018 --- [nio-8080-exec-1] o.a.c.c.C.[Tomcat].[localhost].[/]       : Initializing Spring DispatcherServlet 'dispatcherServlet'
2024-01-22T21:23:17.534+01:00  INFO 5018 --- [nio-8080-exec-1] o.s.web.servlet.DispatcherServlet        : Initializing Servlet 'dispatcherServlet'
2024-01-22T21:23:17.536+01:00  INFO 5018 --- [nio-8080-exec-1] o.s.web.servlet.DispatcherServlet        : Completed initialization in 2 ms
2024-01-22T21:23:34.541+01:00 DEBUG 5018 --- [nio-8080-exec-1] s.s.LnurlAuthSessionAuthenticationFilter : got lnurl-auth session migration request for k1 'd4f067cf72b94baddac1b3856b231669c3239d59cce95ef260da58363fb01822'
2024-01-22T21:23:37.636+01:00 DEBUG 5018 --- [nio-8080-exec-2] s.s.LnurlAuthSessionAuthenticationFilter : got lnurl-auth session migration request for k1 'd4f067cf72b94baddac1b3856b231669c3239d59cce95ef260da58363fb01822'
2024-01-22T21:23:40.715+01:00 DEBUG 5018 --- [nio-8080-exec-4] s.s.LnurlAuthSessionAuthenticationFilter : got lnurl-auth session migration request for k1 'd4f067cf72b94baddac1b3856b231669c3239d59cce95ef260da58363fb01822'
2024-01-22T21:23:43.791+01:00 DEBUG 5018 --- [nio-8080-exec-3] s.s.LnurlAuthSessionAuthenticationFilter : got lnurl-auth session migration request for k1 'd4f067cf72b94baddac1b3856b231669c3239d59cce95ef260da58363fb01822'
2024-01-22T21:23:46.873+01:00 DEBUG 5018 --- [nio-8080-exec-5] s.s.LnurlAuthSessionAuthenticationFilter : got lnurl-auth session migration request for k1 'd4f067cf72b94baddac1b3856b231669c3239d59cce95ef260da58363fb01822'
2024-01-22T21:23:49.960+01:00 DEBUG 5018 --- [nio-8080-exec-7] s.s.LnurlAuthSessionAuthenticationFilter : got lnurl-auth session migration request for k1 'd4f067cf72b94baddac1b3856b231669c3239d59cce95ef260da58363fb01822'
2024-01-22T21:23:53.036+01:00 DEBUG 5018 --- [nio-8080-exec-8] s.s.LnurlAuthSessionAuthenticationFilter : got lnurl-auth session migration request for k1 'd4f067cf72b94baddac1b3856b231669c3239d59cce95ef260da58363fb01822'
2024-01-22T21:23:53.057+01:00 DEBUG 5018 --- [nio-8080-exec-6] .s.w.LnurlAuthWalletAuthenticationFilter : got lnurl-auth wallet authentication request for k1 'd4f067cf72b94baddac1b3856b231669c3239d59cce95ef260da58363fb01822'
2024-01-22T21:23:53.068+01:00 DEBUG 5018 --- [nio-8080-exec-6] org.tbk.lnurl.auth.InMemoryK1Cache       : Remove k1 'org.tbk.lnurl.simple.auth.SimpleK1@5be9b6ad' from in-memory k1 cache: EXPLICIT
2024-01-22T21:23:53.069+01:00 DEBUG 5018 --- [nio-8080-exec-6] rlAuthWalletAuthenticationFailureHandler : Received invalid lnurl-auth request: 03b64a39161d2695add91a89f9c3440a1e4d5b4e055dd19f27208975c002575bcf
```

I don't understand why k1 was removed so quickly but this is not the topic of this issue, i think it would be better to know what was removed from cache instead of `org.tbk.lnurl.simple.auth.SimpleK1@5be9b6ad`

**Comment by theborakompanioni** [comment]
Hey @straumat!

Yes.. `SimpleK1` needs a `toString()` method :raised_hands:
In this case, it was most probably `d4f067cf72b94baddac1b3856b231669c3239d59cce95ef260da58363fb01822`.
Did you try to log in and provided an invalid signature?

**Comment by straumat** [comment]
Yes, indeed, it requires a `toString()` call
I tried to login normally but I dit not yet implement all services like `lnurlAuthSecurityService`... so I was not expecting it to work completely, I just go step by step to understand the different components usage.

**Comment by theborakompanioni** [comment]
> Yes, indeed, it requires a `toString()` call I tried to login normally but I dit not yet implement all services like `lnurlAuthSecurityService`... so I was not expecting it to work completely, I just go step by step to understand the different components usage.

Hey @straumat! I think a key part is the `LnurlAuthPairingService` - you need to create a user from a successful login with a `LinkingKey`, if it does not exist yet (and if that is your use case).
- [`LnurlAuthPairingServiceImpl` in `[spring-lnurl-auth-example-application](https://github.com/theborakompanioni/bitcoin-spring-boot-starter/tree/devel/incubator/spring-lnurl/spring-lnurl-auth-example-application)`](https://github.com/theborakompanioni/bitcoin-spring-boot-starter/blob/devel/incubator/spring-lnurl/spring-lnurl-auth-example-application/src/main/java/org/tbk/lightning/lnurl/example/security/LnurlAuthPairingServiceImpl.java#L29)
or
- [`TestPairingService` in integration tests of `spring-lnurl-auth-security`](https://github.com/theborakompanioni/bitcoin-spring-boot-starter/blob/devel/incubator/spring-lnurl/spring-lnurl-auth-security/src/integTest/java/org/tbk/spring/lnurl/security/test/shared/TestPairingService.java#L31)

You are the first one using it besides myself, so I can surely adapt and make changes according to your use-case!
I have not tried using it in combination with `InMemoryUserDetailsManager`, like you do currently in `royllo` - but maybe I can integrate classes to make this easier.

There is definitely room for improvement, e.g. you do not know the `action` (`register`, `login`, `link`, `auth`) in the PairingService.
So, if you let me know what you think should be improved, that would help a lot :pray:

**Comment by theborakompanioni** [comment]
Hey @straumat!

So the problem apparently is, that [`AbstractTokenAuthenticationProvider#L29`](https://github.com/theborakompanioni/bitcoin-spring-boot-starter/blob/6a9a089638e2e2108d78338d2463a83727a3ad78/incubator/spring-lnurl/spring-lnurl-auth-security/src/main/java/org/tbk/spring/lnurl/security/AbstractTokenAuthenticationProvider.java#L29) tries to retrieve the user. If it does not find an existing one, a `org.springframework.security.core.userdetails.UsernameNotFoundException` with (unfortunately very meaningless) message "<wallet_pubkey>" is thrown.

So, you need to either (a) add a user with username set to the public key of the wallet you want to log in with, or (b) create the user in `LnurlAuthPairingService#pairK1WithLinkingKey`.

Maybe `LnurlAuthPairingService` should get an additional (optional) method `createUserForLinkingKeyIfNecessary` or something like that? This still needs to be thought through properly.

Can you proceed your development with this information?

**Comment by straumat** [comment]
HI! Thanks a lot for your messages. Indeed, i still have things to implement.

I'm going quite slowly because I want to understand each step.

I'm writing an article [here](https://medium.com/@straumat/bitcoin-java-lnurl-auth-with-spring-boot-d2aae3ce2be2) and will publish it once everythings works, the target is a Java/Spring boot developer with no previous knowledge in lighting network.

If you have some time, I have some questions:
- What do you put in the session/cookie? The user id or the k1? if k1, do you sign it or plain K1 ?
- Why do you keep track of the k1 in the database ? Is it because it is used by session/cookie to identify the user when
  he comes back ?
- What do you put the session/cookie when on step1 (cf. medium article), when you generate the qr code ?
- What do you put the session/cookie when on step3 (cf. medium article), when the authentication is valid.
- Why is there a k1cache ? why not generate a new one each time its required ? there is no chance you generate two times the same number ? 
- Why do you for the user to declare a k1Manager ? why the bean is not created by you by default ?

Maybe this issue is not the best place to ask question, tell me what you prefer.
If you don't have time, i would understand

**Comment by theborakompanioni** [comment]
> HI! Thanks a lot for your messages. Indeed, i still have things to implement.
> 
> I'm going quite slowly because I want to understand each step.

:+1: 

> 
> I'm writing an article [here](https://medium.com/@straumat/bitcoin-java-lnurl-auth-with-spring-boot-d2aae3ce2be2) and will publish it once everythings works, the target is a Java/Spring boot developer with no previous knowledge in lighting network.

Unfortunately, I cannot see the blog entry now. Very eager to know what you are writing : -)
> 
> If you have some time, I have some questions:
> 
>     * What do you put in the session/cookie? The user id or the k1? if k1, do you sign it or plain K1 ?

It is a normal spring session cookie created by `spring-session`. See e.g. https://www.javadevjournal.com/spring/spring-session/
I think the default is a base64 encoded UUIDv4 value (to identify and find associated session data).
e.g. `NGUzMjQ5MDAtMTEzNy00MDE4LTliYTMtNzQyNGNkM2YxZDMz` -> `4e324900-1137-4018-9ba3-7424cd3f1d33`

> 
>     * Why do you keep track of the k1 in the database ? Is it because it is used by session/cookie to identify the user when
>       he comes back ?

`k1` does not necessarily need to be tracked in a database. It can be tracked in a key-value store, an internal in-memory cache, or some other way. The reason why `k1` needs to be stored for some time is, that the web login creates a QRCode (or a an lnurl encoded string) for an external device (a lightning wallet or a lightning-login service e.g. [lnpass](https://lnpass.github.io)) to use. `k1` acts as the pairing data. See [LUD-04](https://github.com/lnurl/luds/blob/luds/04.md#server-side-generation-of-auth-url-and-signature-verification) on what is done server-side.

Whenever a wallet provides a valid signature for a `k1` value, the web session will be authenticated.
You can see that when observing the requests made on the login page of the example application:
You will see a lot of `403 FORBIDDEN` requests, until the wallet successfully provides a signature.
As this is just a simple example app it will poll the endpoint, but this can be done via websockets or any other protocol that fits your needs.


> 
>     * What do you put the session/cookie when on step1 (cf. medium article), when you generate the qr code ?

Sorry.. I not sure I understand the question.

The QRCode is generated whenever the user visits the login page. Which is basically just the URL of the service + a newly created `k1` value. 

> 
>     * What do you put the session/cookie when on step3 (cf. medium article), when the authentication is valid.

Same as question 1. it is a spring session cookie (which is just an identifier to later find the session data).
In the example application, session data is stored in an sqlite database. But you can use anything that fits your needs!

> 
>     * Why is there a k1cache ? why not generate a new one each time its required ? there is no chance you generate two times the same number ?

We need to know which web session can be authenticated when a wallet provides a valid signature (+ linking key) for a `k1` value. The value is immediately invalidated when a valid requests arrives or purged after some time (application specific, in the example app it is 5mins I guess). `k1` is a randomly generated 32 bytes of data - very unlikely to create the same number twice (much less within the expiry time - but of course.. not impossible, rather improbable). The example app uses `SecureRandom` - you can use any other RNG.

> 
>     * Why do you for the user to declare a k1Manager ? why the bean is not created by you by default ?

Good question. There is no particular reason why it was done this way.
Take a look at [`K1AutoConfiguration`](https://github.com/theborakompanioni/bitcoin-spring-boot-starter/blob/devel/incubator/spring-lnurl/spring-lnurl-auth-autoconfigure/src/main/java/org/tbk/spring/lnurl/config/K1AutoConfiguration.java).
Currently a user can customize this behaviour by injecting an own `K1Factory` or `K1Cache`.
But I see no downside in annotating the `k1Manager` with `@ConditionalOnMissingBean` as well.. would that benefit your needs?
Of course, you can always refrain from using the `K1AutoConfiguration` all together.. as you can see, the autoconfig class is very simple.
  
> 
> 
> Maybe this issue is not the best place to ask question, tell me what you prefer. If you don't have time, i would understand

No worries, sir. I prefer to answer them here in case someone has the similar questions! All good!
There is nothing more important than helping users.

I hope I could give you helpful answers. Please feel free to suggest improvements or ask any other question that comes to mind :pray:

**Comment by straumat** [comment]
It's strange for the blog posStrange for the article, it says with the link that you should be able to access the draft. It's a work in progress, but
I start to have a good idea of how things work. I may not work a lot on in next week, but I will go back to it in two
weeks.

To make things easier to find, I created a new issue with some questions. You an close this issue as the title may not represent the content of this issue

**Comment by theborakompanioni** [comment]
> To make things easier to find, I created a new issue with some questions. You an close this issue as the title may not represent the content of this issue

Good idea! Closing now as the original issue is fixed in 6a9a089638e2e2108d78338d2463a83727a3ad78. :+1:
Thank you :pray:
